feat: swaps the network Select with a Modal that also shows unsupported chains

hyper54n3 opened the pull request:
[Figma](https://www.figma.com/design/FGmbApkVo2Nb4NH259CAME/Defuse?node-id=4596-14812&t=liXz0HL8uTptHZRF-4)

- [X] Create a new modal for selecting the network
- [X] Swap to the modal in Deposit form
- [X] Swap to the modal in Withdraw form
- [x] add (?) tooltip to Unsupported Networks

Localhost preview:
![image](https://github.com/user-attachments/assets/d23991e3-792f-4958-b580-857f23ba0a08)

hyper54n3 committed:
feat: swaps the modal Select with a Modal that also shows unsupported chains

hyper54n3 committed:
chore: updates based on design feedback, using the same design as the assets modal as a guide

hyper54n3 committed:
fix: placeholder text change

hyper54n3 committed:
feat: adds new network modal to withdraw form
feat: utility to map different types between deposit and withdraw forms

hyper54n3 committed:
feat: add tooltip info to unsupported networks

hyper54n3 committed:
fix: highlight selected network

hyper54n3 committed:
Merge branch 'beta' of github.com:defuse-protocol/defuse-sdk into feat/select-network-modal

hyper54n3 committed:
fix: wrong prop name

jobotics commented on src/components/Modal/ModalSelectNetwork.tsx:
```diff
@@ -0,0 +1,143 @@
+
+        <div className="z-10 flex-1 overflow-y-auto  -mr-[var(--inset-padding-right)] pr-[var(--inset-padding-right)]">
+          {Object.keys({ ...filteredChains, ...disabledChains }).length ===
+          0 ? (
```
You can utilize this in func as it is used in several places, something like `hasChains({ ...filteredChains, ...disabledChains })` and below the same `hasChains(filteredChains)`.

jobotics commented on src/features/deposit/components/DepositForm/index.tsx:
```diff
@@ -185,27 +198,29 @@ export const DepositForm = ({ chainType }: { chainType?: ChainType }) => {
-                />
-              )}
+              render={({ field }) => {
+                const label =
```
What is the point of placing those variables here? Why not move them out of JSX up or place them directly in the component below?

hyper54n3 replied:
Just a bit more readable than straight in the props, but really a question of preference.
I'll update it to have them in the props directly

jobotics commented on src/utils/blockchain.ts:
```diff
@@ -0,0 +1,58 @@
+  blockchain: BlockchainEnum
+): SupportedChainName {
+  // Split the blockchain string by ':' and take the first part
+  const blockchainToChainMap: Record<BlockchainEnum, SupportedChainName> = {
```
Btw, we already have such a map `networkSelectToLabel`, so maybe we can update it to?

rfdlp committed:
fix: feedback from PR review

hyper54n3 committed:
fix: remove unused function

cawabunga-bytes commented on src/utils/blockchain.ts:
```diff
@@ -0,0 +1,52 @@
+    [BlockchainEnum.ZCASH]: "zcash",
+    [BlockchainEnum.GNOSIS]: "gnosis",
+    [BlockchainEnum.BERACHAIN]: "berachain",
+  }
```
I think we already have such function in src/utils/adapters.ts:19

cawabunga-bytes commented on src/features/deposit/components/DepositForm/index.tsx:
```diff
@@ -269,7 +280,12 @@ export const DepositForm = ({ chainType }: { chainType?: ChainType }) => {
 
-function getBlockchainsOptions(): Record<
+export function isAuroraVirtualChain(network: BlockchainEnum): boolean {
+  const virtualChains = [BlockchainEnum.TURBOCHAIN]
```
Aurora is also virtual chain, is it skipped deliberately? I think such info should be stored at some constant config.

Overall current chains configuration is big mess. Chain configuration is evenly distributed across entire codebase. For example, we have PoA Bridge chain id -> internal chain id mapping in one file; aurora engine contract ids in another file; chain labels and icons in 3rd file; evm chain id in 4th file and so on. 

So as for now it would better to reuse existing config (for example, use `siloToSiloAddress` or `auroraEngineContractId` to determine if it is an aurora powered chain). And as a long term it would nice to aggregate chain configs in one place.

cawabunga-bytes commented on src/features/withdraw/components/WithdrawForm/index.tsx:
```diff
@@ -388,28 +410,31 @@ export const WithdrawForm = ({
+                    </Select.Hint>
+                  }
+                  disabled={
+                    blockchainSelectItems &&
```
nit: `blockchainSelectItems` is always defined

cawabunga-bytes commented on src/features/withdraw/components/WithdrawForm/index.tsx:
```diff
@@ -188,11 +195,26 @@ export const WithdrawForm = ({
 
+  const { setModalType, onCloseModal } = useModalStore((state) => state)
+
+  const onChangeNetwork = (network: BlockchainEnum) => {
```
I suggest not using `BlockchainEnum`, because it is specific to PoA bridge, but we operate with other "bridges" too. For example, Aurora Engine chains have own bridge (transfer via silo contract), Near has direct withdrawal (but still we can consider it a "direct" bridge). And Omni bridge is also coming. So I suggest using `SupportedChainName` instead.

cawabunga-bytes commented on src/features/withdraw/components/WithdrawForm/index.tsx:
```diff
@@ -188,11 +195,26 @@ export const WithdrawForm = ({
+      token,
+      selectNetwork: onChangeNetwork,
+      selectedNetwork: chainNameToBlockchain(blockchain),
+    })
```
Personally I don't like such dialogs/modals imperative control. I prefer a declarative approach, where you just render the modal and pass `isOpen` prop to it. 1. It is more conventional to React (and other declarative frameworks). 2. Current implementation hard to use and properly write types. But it's up to you to choose what the best here is.

rfdlp replied:
I agree with you here. Did it this way to follow the existing pattern. Will change it

rfdlp committed:
Merge branch 'beta' of github.com:defuse-protocol/defuse-sdk into feat/select-network-modal

dr-block commented on src/components/Modal/ModalSelectNetwork.tsx:
```diff
@@ -0,0 +1,143 @@
+    return Object.fromEntries(
+      Object.entries(disabledChains).filter(
+        ([key, chain]) =>
+          chain.label.toLowerCase().includes(searchValue.toLowerCase()) ||
```
seems a duplicated logic of filtration, can we make it reusable ?

dr-block commented on src/components/Modal/NetworksList.tsx:
```diff
@@ -0,0 +1,50 @@
+import { PoweredByAuroraLabel } from "../PoweredByAuroraLabel"
+
+interface NetworkListProps {
+  networks: Record<string, { label: string; icon: ReactNode; value: string }>
```
if chain value can be only BlockchainEnum, why we have here string and below BlockchainEnum ?

dr-block commented on src/features/deposit/components/DepositForm/index.tsx:
```diff
@@ -186,24 +199,22 @@ export const DepositForm = ({ chainType }: { chainType?: ChainType }) => {
+                <SelectTriggerLike
+                  label={chainOptions[network ?? ""]?.label ?? "Select network"}
+                  icon={chainOptions[network ?? ""]?.icon ?? <EmptyIcon />}
+                  onClick={() => openModalSelectNetwork()}
```
onClick ={openModalSelectNetwork}

dr-block commented on src/features/withdraw/components/WithdrawForm/index.tsx:
```diff
@@ -388,28 +410,31 @@ export const WithdrawForm = ({
+                  icon={
+                    blockchainSelectItems[field.value]?.icon ?? <EmptyIcon />
+                  }
+                  onClick={() => openModalSelectNetwork()}
```
onClick= openModalSelectNetwork

dr-block commented on src/features/withdraw/components/WithdrawForm/index.tsx:
```diff
@@ -388,28 +410,31 @@ export const WithdrawForm = ({
+                  onClick={() => openModalSelectNetwork()}
+                  hint={
+                    <Select.Hint>
+                      {Object.keys(blockchainSelectItems).length === 1
```
why not reuse Object.keys(blockchainSelectItems).length === 1?

hyper54n3 committed:
fix: change modal to declarative
chore: reuse existing adapter
update: change modal to use SupportedChainName instead of BlockchainEnum
chore: abstract chain filtering function

hyper54n3 committed:
Merge branch 'beta' of github.com:defuse-protocol/defuse-sdk into feat/select-network-modal

hyper54n3 committed:
chore: undo modal changes

vashchukmaksim commented on src/components/Modal/ModalNoResults.tsx:
```diff
@@ -3,15 +3,17 @@ import type { FC } from "react"
+  text = "No assets found",
 }) => {
   return (
     <div className="flex justify-center items-center flex-col gap-2.5">
```
**not related to this PR** but hm... we use props from Radix and classes from Tailwind notation together 🤔

capybara-madness replied:
not related to this PR but hm... we use props from Radix and classes from Tailwind notations together 🤔

vashchukmaksim commented on src/components/Network/ModalSelectNetwork.tsx:
```diff
@@ -0,0 +1,140 @@
+      string,
+      { label: string; icon: ReactNode; value: string }
+    > = {}
+    Object.values(chains).map((chain) => {
```
We don't assign `map` result anywhere but it in fact creates a new array while what we need is a side effect to populate an object. I would change it to `forEach` instead or even better to `.reduce` and avoid assignment to empty object

capybara-madness replied:
We don't assign map result anywhere but it in fact creates a new array while what we need is a side effect to populate an object. I would change it to forEach instead or even better to .reduce and avoid assignment to empty object

vashchukmaksim commented on src/components/Network/ModalSelectNetwork.tsx:
```diff
@@ -0,0 +1,140 @@
+        disabledChains[chain.value] = chain
+      }
+    })
+    return filterChains(disabledChains, searchValue)
```
nit: We kinda do another loop over the same chains again here, can we avoid that?

capybara-madness replied:
nit: We kinda do another loop over the same chains again here, can we avoid that?

vashchukmaksim commented on src/components/Network/ModalSelectNetwork.tsx:
```diff
@@ -0,0 +1,140 @@
+  const availableNetworks = Object.keys(filteredChains).map(
+    (key) => key as BlockchainEnum
+  )
+  const disabledNetworks = Object.keys(disabledChains()).map(
```
Maybe just wrap `disabledChains` is `useMemo` above rather than calling it here on every render

capybara-madness replied:
Maybe just wrap `disabledChains` in `useMemo` above rather than calling it here on every render

vashchukmaksim commented on src/components/Network/ModalSelectNetwork.tsx:
```diff
@@ -0,0 +1,140 @@
+        </div>
+
+        <div className="z-10 flex-1 overflow-y-auto  -mr-[var(--inset-padding-right)] pr-[var(--inset-padding-right)]">
+          {Object.keys({ ...filteredChains, ...disabledChains }).length ===
```
Hm... but `disabledChains` is a function

capybara-madness replied:
Hm... but disabledChains is a function, do not spread it 😇 Actually as for below we can just simplify a check and avoid spreading and creating a new object with it: `availableNetworks.length === 0 && disabledNetworks.length === 0`

vashchukmaksim commented on src/components/Network/ModalSelectNetwork.tsx:
```diff
@@ -0,0 +1,140 @@
+            />
+          ) : (
+            <div className="flex flex-col gap-2 divide-y divide-gray-300">
+              {Object.keys(filteredChains).length > 0 && (
```
Since here we do this check for `NetworkList` and use `availableNetworks` that we've already got using `Object.keys(filteredChains)` above it seems logical check for it directly: `availableNetworks.length > 0`

capybara-madness replied:
Since here we do this check for `NetworkList` and use `availableNetworks` that we've already got using `Object.keys(filteredChains)` above it seems logical check for it directly: `availableNetworks.length > 0`

vashchukmaksim commented on src/components/Network/ModalSelectNetwork.tsx:
```diff
@@ -0,0 +1,140 @@
+                  />
+                </div>
+              )}
+              {Object.keys(disabledNetworks).length > 0 && (
```
Same as for `availableNetworks` above

vashchukmaksim replied:
ook, you actually do check `disabledNetworks` and not `disabledChains` but it's an array, why `Object.keys` then?

vashchukmaksim commented on src/features/deposit/components/DepositForm/index.tsx:
```diff
@@ -80,12 +86,18 @@ export const DepositForm = ({
+  const onCloseNetworkModal = () => setIsNetworkModalOpen(false)
+
+  const onChangeNetwork = (network: SupportedChainName) => {
+    // console.log("onChangeNetwork", network)
```
leftover console.log

capybara-madness replied:
leftover `console.log`

vashchukmaksim commented on src/features/deposit/components/DepositForm/index.tsx:
```diff
@@ -195,25 +210,33 @@ export const DepositForm = ({
+                    onClick={() => setIsNetworkModalOpen(true)}
+                    hint={
+                      <Select.Hint>
+                        {Object.keys(chainOptions).length === 1
```
nit: we iterate twice with `Object.keys` to do length check. Let's maybe move it outside of JSX: `const isOnlyChainOptions = Object.keys(chainOptions).length === 1`

capybara-madness replied:
nit: we iterate twice with `Object.keys` to do length check. Let's maybe move it outside of JSX: `const isOnlyChainOption = Object.keys(chainOptions).length === 1`

vashchukmaksim commented on src/utils/blockchain.ts:
```diff
@@ -0,0 +1,24 @@
+  return Object.fromEntries(
+    Object.entries(candidates).filter(
+      ([key, chain]) =>
+        chain.label.toLowerCase().includes(searchValue.toLowerCase()) ||
```
We do `searchValue.toLowerCase()` three times here (even though sometimes one or two won't be called). Let's do it once after `if` check: `const lowerCaseSearch = searchValue.toLowerCase()` and use it for checks

capybara-madness replied:
We do `searchValue.toLowerCase()` three times here (even though sometimes one or two won't be called). Let's do it once after if check: `const lowerCaseSearch = searchValue.toLowerCase()` and use it for checks

capybara-madness commented on src/components/Network/ModalSelectNetwork.tsx:
```diff
@@ -0,0 +1,140 @@
+                  />
+                </div>
+              )}
+              {Object.keys(disabledNetworks).length > 0 && (
```
ook, you actually do check `disabledNetworks` and not `disabledChains` but it's an array already, why `Object.keys` then?

hyper54n3 committed:
fix: pr feedback

hyper54n3 committed:
Merge branch 'beta' of github.com:defuse-protocol/defuse-sdk into feat/select-network-modal

hyper54n3 committed:
Merge branch 'beta' of github.com:defuse-protocol/defuse-sdk into feat/select-network-modal

hyper54n3 committed:
fix: import for BlockchainEnum type

capybara-madness commented on src/components/Network/ModalSelectNetwork.tsx:
```diff
@@ -0,0 +1,137 @@
+        </div>
+
+        <div className="z-10 flex-1 overflow-y-auto  -mr-[var(--inset-padding-right)] pr-[var(--inset-padding-right)]">
+          {[...availableNetworks, ...disabledNetworks].length === 0 ? (
```
```suggestion
          {availableNetworks.length === 0 && disabledNetworks.length === 0 ? (
```

- not creating a new array
- constant time complexity

amdefuse commented:
:tada: This PR is included in version 1.0.0-beta.115 :tada:

The release is available on:
- [GitHub release](https://github.com/defuse-protocol/defuse-sdk/releases/tag/v1.0.0-beta.115)
- [npm package (@beta dist-tag)](https://www.npmjs.com/package/@defuse-protocol/defuse-sdk/v/1.0.0-beta.115)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
